EZP-29290: As an Editor, I want to have preconfigured "Author" section while creating the content

konradoboza opened the pull request:
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-29290](https://jira.ez.no/browse/EZP-29290)
| **Bug/Improvement**| no
| **New feature**    | yes
| **Target version** | `master`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | yes

JIRA excerpt:
>Currently, in v2 there is no preconfigured author section (fields filled in with the name and e-mail of the currently logged user). It would be more convenient for the editors to have this fields set automatically while creating a new content. Also, prefilling section with the default data should be configurable at the Author FieldType level.

This is the kernel implementation of this feature. It is followed by the changes in `repository-forms` and `admin-ui-bundle` which will enable a user to choose whether Author FieldType should be prefilled by default from the ContentType creation view.

**Is required by:**
- ezsystems/repository-forms/pull/251
- ezsystems/ezplatform-admin-ui/pull/586

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

gggeek commented:
I'd say default for the fieldType should be Empty, not current-user.

In fact in more than 10 years using eZ, I have never seen anyone using an author field to store the current user. That info is embedded into content metadata anyway. Whenever an author field is used, it is because the persone logged ion to input thr data is _not_ the author that has to be shown as content creator

konradoboza commented:
@gggeek thank you for the input. Default option could be easily changed, that's not a problem at all. Prefilling author FieldType with current user's data was present in eZ Publish and eZ Platform v1 so as a matter of consistency we would like to have it also in v2. We also aim to give the Editor possibility to keep current behavior as well.

konradoboza commented:
Few things I'm not sure about:
- is injecting `Repository` into `AuthorConverter` the best way to solve it? Tests got more complicated, maybe I'm missing something.
- test is most likely failing because of this https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/API/Repository/Tests/FieldType/BaseIntegrationTest.php#L1038. when we decide to have author data prefilled by default this array won't be empty. How should I fix it?
- should logic responsible for creating mocks which is duplicated in `SPI/Tests/FieldType/AuthorIntegrationTest.php` and `Core/Persistence/Legacy/Tests/Content/FieldValue/Converter/AuthorTest.php` be moved to e.g. Trait?
Any advice would be much appreciated :) @andrerom @alongosz

andrerom commented:
> is injecting Repository into AuthorConverter the best way to solve it? Tests got more complicated, maybe I'm missing something.

That is probably not the way to do it, as Converter is performed down in Storage Engine and should now know about repository, besides circular references.

As for how and the other question maybe @alongosz and @bdunogier can give you some input.

konradoboza commented:
I changed approach towards implementing this feature in `repository-forms`. Followup PR's need to wait for https://github.com/ezsystems/repository-forms/pull/249 as it requires further changes.

andrerom commented on eZ/Publish/Core/FieldType/Author/Type.php:
```diff
@@ -21,6 +22,23 @@
+    /**
+     * Default value empty.
+     */
+    const DEFAULT_EMPTY = -1;
```
Any reason why you picked -1 over for instance 0? Other cases you found like that?
Asking as I can't right now remember us using negative ints much for int fields.

konradoboza replied:
I was having hard times to differentiate this value from this which I got from the database for existing FieldTypes (e.g. Article), as we wanted to have prefilling by default there.

andrerom replied:
Ok, so you're not using 0 as todays default (Null) for data_int1. is casted to int in Storage engine so it becomes 0, which is causing your trouble to decide on value. Right?

Is 0 mapped to DEFAULT_CURRENT_USER somewhere for us btw, so editing existing data works as intended?

@alongosz SIDE: you see any issues with using negative int's for data_int1 column? Or should we rather keep the values in the positive range?

konradoboza replied:
>Ok, so you're not using 0 as todays default (Null) for data_int1. is casted to int in Storage engine so it becomes 0, which is causing your trouble to decide on value. Right?

Yes, that was my point.

>Is 0 mapped to DEFAULT_CURRENT_USER somewhere for us btw, so editing existing data works as intended?

So you are suggesting switching values? (0, both for prefilling chosen by editor and also for no value from db, 1 for no-prefilling option). It could be confusing, but probably would work.

andrerom replied:
> So you are suggesting switching values?

not necessarily, before even suggesting that I was asking how the system behaves when db has null value. Both when it comes to content type view/edit and for content edit.

context: People upgrading would expect default value to just be current user somehow, so if code here handles that or if we are missing upgrade script where those upgrading can choose between setting these to 1 for user or -1/2 for todays null value.

konradoboza replied:
@andrerom I tested adding/editing content/content type with/without prefilling, everything seems to be working fine.

andrerom commented:
@konradoboza I'm wondering if this phase 2 on kernel and here should rather go to master as it's a new feature and not requested by customer so we don't need to rush it. would that work for you?

konradoboza commented:
@andrerom sure, it is fine with me. So this one should be rebased to master, how about other two?

andrerom commented:
@konradoboza so which one depends on which? Which PR's are "phase 1", aka minimal scope, and which once are followups to add config in field type?

I's the followups I'm suggesting we take to master basically, leaving phase 1 for the shortly upcoming 2.2.2 release.

konradoboza commented:
@andrerom all three PR's should be treated as one feature, so this is "phase 2", I updated descriptions for https://github.com/ezsystems/repository-forms/pull/251 and https://github.com/ezsystems/ezplatform-admin-ui/pull/586 as they depend on kernel.

"Phase 1" was already merged and was about having prefilled Author FieldType, without having a config. Ref: https://github.com/ezsystems/repository-forms/pull/249

andrerom commented:
Ok, clear. Then go for master on the remaining PR's, should not have any conflicts so hopefully simple rebase ;)

andrerom commented:
Actually, we can try to just change target on PR's, if it does not conflict we are good!

konradoboza commented:
Fixed checking if any value came from the database in order to have a default selected radio in the Content Type edit view for existing ones. https://github.com/ezsystems/ezpublish-kernel/pull/2392/commits/9a91b9c18113c470aa049f5f85f169f4afc001fc#diff-564956cdbfd9b8b86cfb41dc3ddbf243R79

andrerom commented on eZ/Publish/Core/REST/Common/Tests/FieldTypeProcessor/AuthorProcessorTest.php:
```diff
@@ -0,0 +1,69 @@
+        return array_map(
+            function ($constantName) {
+                return array(
+                    array('defaultAuthor' => $constantName),
```
[]

Aka use `[]` in new code and lines that are changed, and leave `array` in places you don't change to unnecessary branch conflicts.

andrerom commented on eZ/Publish/SPI/Tests/FieldType/AuthorIntegrationTest.php:
```diff
@@ -98,17 +109,17 @@ public function getFieldDefinitionData()
         return new Content\FieldValue(
-            array(
-                'data' => array(
-                    array(
```
can skip these changes

andrerom commented on eZ/Publish/Core/Persistence/Legacy/Content/FieldValue/Converter/AuthorConverter.php:
```diff
@@ -123,20 +131,20 @@ private function generateXmlString(array $authorValue)
      * @param string $xmlString XML String stored in storage engine
      *
-     * @return \eZ\Publish\Core\FieldType\Author\Value
+     * @return array
```
`\eZ\Publish\Core\FieldType\Author\Value[]`

andrerom commented on eZ/Publish/Core/Persistence/Legacy/Content/FieldValue/Converter/AuthorConverter.php:
```diff
@@ -123,20 +131,20 @@ private function generateXmlString(array $authorValue)
     {
         $dom = new DOMDocument('1.0', 'utf-8');
-        $authors = array();
+        $authors = [];
```
can skip these changes

konradoboza commented:
I reverted all the unnecessary changes to array short notation. When there were `[]` and `array()` in one file I changed these to `array()` to be consistent with the previous versions and with the other tests.

konradoboza commented on eZ/Publish/Core/FieldType/Author/Type.php:
```diff
@@ -148,4 +166,51 @@ public function isSearchable()
+            }
+
+            switch ($name) {
+                case 'defaultAuthor':
```
The question is whether we want to be consistent with the other implementations or improve things when it comes to FieldTypes, https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/FieldType/Date/Type.php#L224.

alongosz replied:
Each Field Type can be treated a separate, so implementations can differ. What we need to avoid is switching code style approaches between valid alternatives. For example pushing towards short array syntax was my mistake, it should be done for entire codebase at once.

In this case I find Field Type implementations quite confusing, so I would rather improve.

That said, maybe just extract contents of this `case` block to a separate method? It keeps somewhat consistency and improves readability.

konradoboza commented on eZ/Publish/Core/FieldType/Author/Type.php:
```diff
@@ -148,4 +166,51 @@ public function isSearchable()
+                    ];
+                    if (!in_array($value, $definedTypes, true)) {
+                        $validationErrors[] = new ValidationError(
+                            "Setting '%setting%' is of unknown type",
```
Yes, it is actually my bad, `$defaultValues` sounds much better.

konradoboza commented on eZ/Publish/Core/FieldType/Author/Type.php:
```diff
@@ -148,4 +166,51 @@ public function isSearchable()
+    /**
+     * Validates the fieldSettings of a FieldDefinitionCreateStruct or FieldDefinitionUpdateStruct.
+     *
+     * @param mixed $fieldSettings
```
The question is: should it be changed for every `Type` class? As I see `mixed` is used in all FieldTypes which have settings.

alongosz replied:
No, let's not get out of scope here. It is indeed inconsistent but let's not duplicate past mistakes.

konradoboza commented on eZ/Publish/Core/Persistence/Legacy/Tests/Content/FieldValue/Converter/AuthorTest.php:
```diff
@@ -22,6 +27,7 @@
  */
 class AuthorTest extends TestCase
 {
+    const ADMINISTRATOR_USER_ID = 14;
```
Great catch! I must have forgotten to remove it alongside with the mocks from the previous implementation.

konradoboza commented:
Suggestions from @alongosz applied. Regarding BC: in v1 and Legacy we had prefilled Author data by default. Within the scope of this (and related) PR we give editors an option in CT creation view. Existing CT's (e.g. article) will be prefilled by default. /cc @lserwatka 

![zrzut ekranu 2018-08-08 o 16 18 31](https://user-images.githubusercontent.com/34310128/43842985-c41f133a-9b26-11e8-9aa1-09d843a63fa4.png)

webhdx commented on eZ/Bundle/EzPublishCoreBundle/Resources/views/fielddefinition_settings.html.twig:
```diff
@@ -267,7 +267,16 @@
+    {% endif %}
+    {{ block( 'settings_defaultvalue' ) }}
+</ul>
+{% endblock %}
```
It looks okay.

lserwatka commented:
@barbaragr let's proceed with QA here.

alongosz commented:
Code looks now good.

If we want to merge a BC break, at least please add a note in `doc/bc` stating that since new version, Author Field Type will be prefilled with current user by default. 

I'm against making BC breaks in places we have a choice (e.g. not forced by bugs), but since this is what PMs want, BC note is the least we could do.

konradoboza commented:
BC break note added in: https://github.com/ezsystems/ezpublish-kernel/pull/2411
